### PR TITLE
implement draft-25

### PR DESCRIPTION
**Pull request by kazuho**
* transport:
  * [x] HANDSHAKE_DONE
  * [x] Retry with integrity check
  * [x] correct overflow in packet number decode algorithm
  * [x] Allow use of CRYPTO_BUFFER_EXCEEDED
  * [x] treat invalid CONNECTION_CLOSE as an invalid frame
  * [x] throttle when sending CONNECTION_CLOSE (to be fixed in #256)
  * [x] APPLICATION_CLOSE only in 0-RTT / 1-RTT
  * [ ] ~~send while blocked to avoid idle timeout~~ (separately tracked in #259)
  * [x] error codes for invalid frames
  * [x] symmetric idle timeout
* tls: none applicable
* recovery:
  * [x] ~~PN-space specific PTO (maybe fixed in #253)~~. This PR adopts #257 which is the alternative to #253. Quicly will send an ACK-eliciting packet at every PN space when PTO fires. Doing so is fine for 1-RTT, and there would be no negative impact in practice during handshake too, because with the introduction of HANDSHAKE_DONE, the protocol drops PN spaces as early as possible. @janaiyengar Does that make sense? We need to consider about SPTO too.

### CODE REVIEW

**Comment by kazuho** [comment]
@janaiyengar Please review.

I think the most controversial change is the PTO thing (which you can separately review in #257), but the CI relies on #252 being resolved in certain ways, and I think #257 is more correct than #253 in sense that it sends PINGs in all PN spaces and also simpler.

File: lib/quicly.c [hunk 1]
```diff
@@ -4563,36 +4623,28 @@ int quicly_receive(quicly_conn_t *conn, struct sockaddr *dest_addr, struct socka
...
-                    ret = PTLS_ERROR_NO_MEMORY;
+        if (quicly_is_client(conn)) {
+            /* Running as a client.
+             * Respect "disable_migration" TP sent by the peer at the end of the TLS handshake. */
+            if (conn->super.host.address.sa.sa_family == AF_UNSPEC && dest_addr != NULL && dest_addr->sa_family != AF_UNSPEC &&
+                ptls_handshake_is_complete(conn->crypto.tls) && conn->super.peer.transport_params.disable_active_migration)
+                set_address(&conn->super.host.address, dest_addr);
+        } else {
+            /* Running as a server.
+             * When invoked for the first time, drop handshake context, schedule the first emission of HANDSHAKE_DONE frame. */
```
Comment (janaiyengar) [inline lib/quicly.c]:
```suggestion
             * If handshake was just completed, drop handshake context and schedule the first emission of HANDSHAKE_DONE frame. */
```

**Comment by janaiyengar** [comment]
Thank you for doing this !

**Comment by janaiyengar** [comment]
Regarding SPTO, we need to re-think that a bit I think... maybe do it only on 1-RTT packets. But I don't think we need to delay this PR for that.

File: lib/quicly.c [hunk 2]
```diff
@@ -2215,6 +2200,14 @@ static int on_ack_streams_blocked(quicly_conn_t *conn, const quicly_sent_packet_
     return 0;
 }
 
+static int on_ack_handshake_done(quicly_conn_t *conn, const quicly_sent_packet_t *packet, quicly_sent_t *sent,
+                                 quicly_sentmap_event_t event)
+{
+    if (event == QUICLY_SENTMAP_EVENT_LOST)
```
[the remark this answers is not in the document]
Comment (kazuho) [reply lib/quicly.c]: Thank you for the suggestion. Addressed in 5baf698.

File: lib/quicly.c (hunk 1 above)
Comment (kazuho) [reply lib/quicly.c]: Thank you for the suggestion. Addressed in 5baf698.

**Pull request merged**